presto-hive-common/src/main/java/com/facebook/presto/hive/zorder/ZOrderCurveSearcher.java

```diff
@@ -0,0 +1,249 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.hive.zorder;
+
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.LinkedList;
+import java.util.List;
+
+import static java.lang.String.format;
+
+/**
+ * This class contains functions to search for addresses within given value ranges.
+ */
+public class ZOrderCurveSearcher
+{
+    private final int[] dimensions;
+
+    private ZOrder zOrder;
+
+    private List<Integer> minimumValues;
+    private List<Integer> maximumValues;
+
+    private boolean returnLong;
+    private LinkedList<ZAddressRange> addressRanges;
+
+    /**
+     * Class constructor specifying the ZOrder class and initializing <code>dimensions</code>.
+     * <p/>
+     * The <code>dimensions</code> array stores how many values are stored at each level of the Z-curve.
+     */
+    public ZOrderCurveSearcher(ZOrder zOrder)
```

Comment:
I do not complain that we separate this class out from ZOrder class. One question that we have to ask: What kind of APIs do we want to expose to users? In this case, the code for users will be like:

`Zorder zOrder = new ZOrder(...); ZOrderCurveSearcher searcher = new ZOrderCurveSearcher(zOrder); ....`.

We can see that: (1) we need expose two classes to the user, and (2) the user has to understand their relationship (who depends on who). To make it simple, the user should only be exposed to the follow APIs:

`ZOrder zOrder = new ZOrder(...); zOrder.encode(); .zOrder.decode(...); zOrder.search(...);` 

To achieve this design, basically we want to achieve "friend" relationships between ZOrder and ZOrderCurveSearcher. How ? See if we can use the little trick: https://stackoverflow.com/a/18634125